### PR TITLE
[001] Add test, add crud logic, add liquibase scripts

### PR DESCRIPTION
**Pull request by Yurchenko-Anton**
Add tests for admin service, initialize db via liquibase changeset.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -22,13 +22,29 @@ dependencies {
...
+	testImplementation 'org.hamcrest:hamcrest-all:1.3'
+	testImplementation 'org.projectlombok:lombok:1.18.22'
 	compileOnly 'org.projectlombok:lombok'
 	annotationProcessor 'org.projectlombok:lombok'
 	testImplementation 'org.springframework.boot:spring-boot-starter-test'
 	testImplementation 'org.springframework.security:spring-security-test'
-	implementation group: 'com.mysql', name: 'mysql-connector-j', version: '8.0.31'
+	implementation group: 'mysql', name: 'mysql-connector-java', version: '8.0.20'
 	implementation group: 'io.jsonwebtoken', name: 'jjwt-api', version: '0.11.5'
-//	implementation 'org.liquibase:liquibase-core'
```
Comment (asinchuk) [inline build.gradle:31]: please remove unused dependency

File: src/main/java/com/example/security/SecurityApplication.java [hunk 2]
```diff
@@ -2,8 +2,10 @@
 
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.cloud.openfeign.EnableFeignClients;
 
 @SpringBootApplication
+@EnableFeignClients
```
Comment (asinchuk) [inline src/main/java/com/example/security/SecurityApplication.java:8]: Do we need this annotation? Please chect if test work without it

File: src/main/java/com/example/security/dto/AuthenticationRequestDTO.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+import lombok.Value;
+
+@Value
+public class AuthenticationRequestDTO {
+
+    String phone;
+    String password;
+
+
```
Comment (asinchuk) [inline src/main/java/com/example/security/dto/AuthenticationRequestDTO.java:11]: please remove empty lines 10&11.

File: src/main/java/com/example/security/dto/ChangeRoleDTO.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+import com.example.security.model.Role;
+import lombok.Value;
+
+@Value
+public class ChangeRoleDTO {
+
+    Long id;
+    Role role;
+
```
Comment (asinchuk) [inline src/main/java/com/example/security/dto/ChangeRoleDTO.java:11]: please remove empty line

File: src/main/java/com/example/security/security/JwtTokenProvider.java [hunk 5]
```diff
@@ -18,30 +18,30 @@
...
     }
 
-    public String createToken(String username, String role) {
+    public String createToken(String username, String role, Long id) {
         Claims claims = Jwts.claims().setSubject(username);
         claims.put("role", role);
+        claims.put("id", id);
         Date now = new Date();
-        Date validity = new Date(now.getTime() + validityInMilli * 1000);
+        Date validity = new Date(now.getTime() + validityInMillis * 1000);
```
Comment (asinchuk) [inline src/main/java/com/example/security/security/JwtTokenProvider.java:44]: please remove multiplier from validity calulation

File: src/main/java/com/example/security/services/AdminService.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    UserRepository userRepository;
+
+    public Optional<User> findUsers(Long id) {
+        return userRepository.findById(id);
+
+    }
+
+    public ResponseEntity<User> changeRole(Long id, Role role) {
+        Optional<User> user = findUsers(id);
+        if (user.isPresent()) {
```
Comment (asinchuk) [inline src/main/java/com/example/security/services/AdminService.java:26]: please refactor to optional stream approach

File: src/main/java/com/example/security/services/AdminService.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            userRepository.save(user.get());
+            return ResponseEntity.ok(user.get());
+        } else {
+            return ResponseEntity.notFound().build();
+        }
+    }
+
+    public ResponseEntity<User> approve(Long id) {
+        Optional<User> user = findUsers(id);
+        if (user.isPresent()) {
```
Comment (asinchuk) [inline src/main/java/com/example/security/services/AdminService.java:37]: please refactor to optional stream approach

File: src/main/java/com/example/security/services/AuthenticationService.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+public class AuthenticationService {
+
+    private final AuthenticationManager authenticationManager;
+    private UserRepository userRepository;
+    private JwtTokenProvider jwtTokenProvider;
+
+
+    public Map<Object, Object> authenticate(AuthenticationRequestDTO request) throws AuthenticationException {
+        return buildResponse(
+                findUser(authenticationManager.authenticate(new UsernamePasswordAuthenticationToken(request.getPhone(), request.getPassword())).isAuthenticated()
```
Comment (asinchuk) [inline src/main/java/com/example/security/services/AuthenticationService.java:28]: method call as parameter is bad practice, please avoid it

File: src/main/java/com/example/security/SecurityApplication.java (hunk 2 above)
Comment (Yurchenko-Anton) [reply src/main/java/com/example/security/SecurityApplication.java:8]: We need this annotation.

Comment (asinchuk) [reply src/main/java/com/example/security/SecurityApplication.java:8]: but we have same annotation specific to test class, please check if test will work if @EnableFeignClients will be removed from application main class

File: src/main/java/com/example/security/config/SecurityConfig.java [hunk 9]
```diff
@@ -41,19 +34,22 @@ protected void configure(HttpSecurity http) throws Exception {
...
     @Bean
     @Override
-    public AuthenticationManager authenticationManagerBean() throws Exception{
+    public AuthenticationManager authenticationManagerBean() throws Exception {
         return super.authenticationManagerBean();
     }
+
+    @Bean
+    protected PasswordEncoder passwordEncoder() {
+        int encryptStrength = 12;
```
Comment (asinchuk) [inline src/main/java/com/example/security/config/SecurityConfig.java:52]: please move this magic number to well named constant

File: src/main/java/com/example/security/entity/User.java [hunk 10]
```diff
@@ -13,10 +13,16 @@ public class User {
...
     @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
-
+    @Column(name = "first_name")
     private String firstName;
+    @Column(name = "last_name")
     private String lastName;
+
+    private String phone;
     private String password;
```
Comment (asinchuk) [inline src/main/java/com/example/security/entity/User.java:22]: Please be consistent in column declaration

File: src/main/java/com/example/security/entity/User.java [hunk 11]
```diff
@@ -13,10 +13,16 @@ public class User {
...
     @Enumerated(value = EnumType.STRING)
     private Role role;
+    @Enumerated(value = EnumType.STRING)
+    private Status status;
+
```
Comment (asinchuk) [inline src/main/java/com/example/security/entity/User.java:27]: just a beautify note - remove last string before curly brace and after it

File: src/main/java/com/example/security/security/JwtTokenProvider.java [hunk 12]
```diff
@@ -56,24 +58,24 @@ public boolean validateToken(String token) throws JwtAuthenticationException {
...
     public String getUsername(String token) {
-        return Jwts.parser()
-                .setSigningKey(secretKey)
-                .parseClaimsJws(token)
-                .getBody()
-                .getSubject();
+        return Jwts.parser().setSigningKey(secretKey).parseClaimsJws(token).getBody().getSubject();
+    }
+
+    public String resolveToken(HttpServletRequest request) {
```
Comment (asinchuk) [inline src/main/java/com/example/security/security/JwtTokenProvider.java:74]: please check if this method required, if it used in one place, refactor it to method variable

File: src/main/java/com/example/security/security/UserDetailsServiceImpl.java [hunk 13]
```diff
@@ -8,17 +8,16 @@
...
 public class UserDetailsServiceImpl implements UserDetailsService {
 
     private final UserRepository userRepository;
 
 
-
     @Override
-    public UserDetails loadUserByUsername(String firstName) throws UsernameNotFoundException {
-        User user = userRepository.findByFirstName(firstName).orElseThrow(()-> new UsernameNotFoundException("User not found"));
+    public UserDetails loadUserByUsername(String phone) throws UsernameNotFoundException {
```
Comment (asinchuk) [inline src/main/java/com/example/security/security/UserDetailsServiceImpl.java:19]: please change to retrieve user by userId

File: src/main/java/com/example/security/services/AuthenticationService.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+@Service
+@AllArgsConstructor
+public class AuthenticationService {
+
+    private final AuthenticationManager authenticationManager;
+    private UserRepository userRepository;
+    private JwtTokenProvider jwtTokenProvider;
+
+
+    public Map<Object, Object> authenticate(AuthenticationRequestDTO request) throws AuthenticationException {
```
Comment (asinchuk) [inline src/main/java/com/example/security/services/AuthenticationService.java:29]: Please check if Map<Object, Object> is reqired

File: src/main/java/com/example/security/services/UsersService.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Optional;
+
+@Service
+@AllArgsConstructor
+public class UsersService {
+    private final int ENCRYPT_STRENGTH = 12;
```
Comment (asinchuk) [inline src/main/java/com/example/security/services/UsersService.java:19]: private static final

File: src/main/java/com/example/security/services/UsersService.java [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+    UserRepository userRepository;
```
Comment (asinchuk) [inline src/main/java/com/example/security/services/UsersService.java:21]: private final

File: src/main/java/com/example/security/services/UsersService.java [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    }
+    public void changePassword(String token, String password) {
+        Long id = Long.parseLong(jwtTokenProvider.decodeToken(token).get("id").toString());
+        User user = userRepository.findById(id)
+                .orElseThrow(() -> new UsernameNotFoundException("user doesn't exist"));
+        user.setPassword(new BCryptPasswordEncoder(ENCRYPT_STRENGTH).encode(password));
+        saveUser(user);
+    }
+
+    public Status validate(User user) {
```
Comment (asinchuk) [inline src/main/java/com/example/security/services/UsersService.java:35]: please check if this method could be private and arrange methods by call order, from top to bottom

File: src/main/java/com/example/security/services/UsersService.java [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+@AllArgsConstructor
+public class UsersService {
+    private final int ENCRYPT_STRENGTH = 12;
+
+    UserRepository userRepository;
+    JwtTokenProvider jwtTokenProvider;
+
+    private void saveUser(User user){
+        userRepository.save(user);
+    }
```
Comment (asinchuk) [inline src/main/java/com/example/security/services/UsersService.java:26]: new string

File: src/main/resources/application.properties [hunk 19]
```diff
@@ -1,10 +1,10 @@
...
 spring.datasource.username=root
-spring.datasource.password=my-secret-pw
+spring.datasource.password=root
 spring.jpa.hibernate.ddl-auto=none
 
 jwt.header=Authorization
-jwt.secret=proselyte
-jwt.expiration=604800
-#spring.liquibase.change-log=classpath:db/changelog/db.changelog-master.xml
+jwt.secret=proselyteproselyteproselyteproselyte
```
Comment (asinchuk) [inline src/main/resources/application.properties:8]: camelCasePlease

File: src/test/java/com/example/security/SecurityApplicationTests.java [hunk 20]
```diff
@@ -1,13 +1,117 @@
...
 
-@SpringBootTest
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.NONE,
+        classes = SecurityApplication.class)
+@AutoConfigureMockMvc
+@TestPropertySource(locations = "classpath:application-test.properties")
 class SecurityApplicationTests {
```
Comment (asinchuk) [inline src/test/java/com/example/security/SecurityApplicationTests.java:33]: please remove this integration tests for now, in future we will add unit tests

File: src/main/java/com/example/security/web/AdminsController.java [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+import com.example.security.services.AdminService;
+import lombok.AllArgsConstructor;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/v1/admins")
```
Comment (asinchuk) [inline src/main/java/com/example/security/web/AdminsController.java:15]: please remove explicit versioning i.e. /auth/admins

File: src/main/java/com/example/security/web/AdminsController.java [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+@PreAuthorize("hasAuthority('users:write')")
+@AllArgsConstructor
+public class AdminsController {
+
+    UserRepository userRepository;
+    AdminService adminService;
+
+    @GetMapping("/get")
```
Comment (asinchuk) [inline src/main/java/com/example/security/web/AdminsController.java:23]: Please separate by request methods not by URLs

File: src/main/java/com/example/security/security/UserDetailsServiceImpl.java [hunk 23]
```diff
@@ -8,17 +8,15 @@
...
 public class UserDetailsServiceImpl implements UserDetailsService {
 
     private final UserRepository userRepository;
 
-
-
     @Override
-    public UserDetails loadUserByUsername(String firstName) throws UsernameNotFoundException {
-        User user = userRepository.findByFirstName(firstName).orElseThrow(()-> new UsernameNotFoundException("User not found"));
+    public UserDetails loadUserByUsername(String id) throws UsernameNotFoundException { // cast to Long
```
Comment (asinchuk) [inline src/main/java/com/example/security/security/UserDetailsServiceImpl.java:18]: Better naming is findByUserName

**Review by asinchuk — APPROVED** [review]

**Comment by asinchuk** [comment]
Ping me before merge

**Pull request merged**